### PR TITLE
[Common] Moved framework agnostic THD kernels to common.

### PR DESCRIPTION
**Pull request by mgoldfarb-nvidia**
# Description

Moves several kernels that are used for context parallelism + THD fused attention to shared location so they can be called by Jax and other frameworks.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [X] Code refractor

## Changes

Moves common code out of pytorch filder.

# Checklist:

- [X] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [X] The functionality is complete
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/common/fused_attn/thd_utils.h [hunk 1]
```diff
@@ -0,0 +1,222 @@
+/*************************************************************************
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See LICENSE for license information.
+ ************************************************************************/
+
+#ifndef TRANSFORMER_ENGINE_FUSED_ATTN_THD_UTILS_H_
+#define TRANSFORMER_ENGINE_FUSED_ATTN_THS_UTILS_H_
```
Comment (xrennvidia) [inline transformer_engine/common/fused_attn/thd_utils.h:8]: THS is a typo at here, right?

**Review by xrennvidia — COMMENTED** [review]
LGTM. Thanks.

Please make sure PyTorch CP tests (L1 tests) passed. Thanks.

**Comment by xrennvidia** [comment]
FYI, it seems like pipeline `20535774` failed, not sure why it shows it completed successfully.

**Comment by zlsh80826** [comment]
Probably needs to add `thd_utils.cu` to the [CMakeLists.txt](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/common/CMakeLists.txt#L64) here.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -70,6 +70,7 @@ jobs:
         run: pip install . -v
         env:
           NVTE_FRAMEWORK: jax
+          MAX_JOBS: 1
```
Comment (mgoldfarb-nvidia) [inline .github/workflows/build.yml:73]: cc @timmoon10

**Review by xrennvidia — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**